File: Baekjoon/dfs/02.알파벳_baekjoon1987.cpp
#include <bits/stdc++.h>
using namespace std;

int r,c, result = 1;
int board[20][20] = {0,};
bool visited[26];
int dx[4] = {0, 1, 0, -1}, dy[4] = {1, 0, -1, 0};

void dfs(int x, int y, int cnt){
  bool end = true;
  
  for(int i=0; i<4; i++){
    int nx = x + dx[i];
    int ny = y + dy[i];
    
    if(nx<0 || nx>=r || ny<0 || ny>=c) continue;
    if(visited[board[nx][ny]]) continue;
    visited[board[nx][ny]] = true;
    end = false;
    dfs(nx, ny, cnt+1);
    visited[board[nx][ny]] = false;
  }
  
  if(end)
    result = max(result, cnt);
}
  
int main(){
  ios::sync_with_stdio(0);
  cin.tie(0); cout.tie(0);
  
  cin>>r>>c;
  
  for(int i=0; i<r; i++){
    for(int j=0; j<c; j++){
      char c;
      cin>>c;
      board[i][j] = (int)c- 65;
    }
  } 
  visited[board[0][0]] = true;
  dfs(0, 0, 1);
  cout<<result;
}

File: Baekjoon/dfs/02.알파벳_baekjoon1987.java
import java.io.*;
import java.util.*;

class Main {
	static int r,c, result = 0;
	static int[] dx = {0,1,0,-1}, dy = {1,0,-1,0};
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");

		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());
		char[][] map = new char[r][c];
		boolean[] visited = new boolean[26];
		for(int i=0; i<r; i++) {
			map[i] = br.readLine().toCharArray();
		}
		visited[map[0][0]-'A'] = true;
		dfs(0, 0, 1, map, visited);
		br.close();
		System.out.print(result);
	}

	static void dfs(int x, int y, int cnt, char[][] map, boolean[] visited){
		boolean end = true;

		for(int i=0; i<4; i++){
			int nx = x + dx[i];
			int ny = y + dy[i];

			if(nx<0 || nx>=r || ny<0 || ny>=c || visited[map[nx][ny]-'A']) continue;
			visited[map[nx][ny]-'A'] = true;
			end = false;
			dfs(nx, ny, cnt+1, map, visited);
			visited[map[nx][ny]-'A'] = false;
		}

		if(end) result = Math.max(result, cnt);
	}
}